Overhaul launch-game.sh

pchote opened the pull request:
Another step towards being able to remove the mod launcher:

- Adds mod chooser popup if zenity available
- Adds at least OSX and hopefully general POSIX support
- Improved text-mode output when zenity not available.

I have used http://www.shellcheck.net/ to check for portability and confirmed that this works as expected under linux (with gnome 3) and macOS.

ltem commented:
It works also under linux with XFCE. 

However if I cancel the Zenity window it will still still start with the `ra`-mod. I would expect that it stops the start procedure or is this on purpose?

pchote committed:
Overhaul launch-game.sh

- Adds macOS support
- Adds mod chooser dialog if zenity available
- Improved text-mode output when zenity not available.

pchote commented:
Good catch!  Fixed.

ltem commented:
It works as expected but if I run it without a mod I get this message. 
```zsh
$ ./launch-game.sh
/usr/bin/zenity
Gtk-Message: GtkDialog mapped without a transient parent. This is discouraged.
```
StackOverflow has a post about it [Link](http://stackoverflow.com/questions/29883211/gtkdialog-mapped-without-a-transient-parent#29883565), but no idea if this is wanted.

obrakmann commented on launch-game.sh:
```diff
@@ -1,9 +1,36 @@
-# Note: this relies on the non-standard -f flag implemented by gnu readlink
-mono OpenRA.Game.exe Engine.LaunchPath="$(readlink -f $0)" "$@"
-if [ $? != 0 -a $? != 1 ]
+MODLAUNCHER=$(python -c "import os; print(os.path.realpath('$0'))")
```
Not a fan of adding another dependency here for this rather mundane task.

pchote replied:
gnu readlink is *already* an extra dependency, and doesn't work on OSX or other systems that use BSD readlink.

obrakmann commented on launch-game.sh:
```diff
@@ -1,9 +1,36 @@
+		if [ "$TITLE" = "cancel" ]; then exit 0
+		elif [ "$TITLE" = "Tiberian Dawn" ]; then MODARG='Game.Mod=cnc'
+		elif [ "$TITLE" = "Dune 2000" ]; then MODARG='Game.Mod=d2k'
+		elif [ "$TITLE" = "Tiberian Sun" ]; then MODARG='Game.Mod=ts'
```
Do we want to add that already?

pchote replied:
This script is for launching the game on a dev/self-compiled version of OpenRA.  That includes TS, so why wouldn't we want to add this?

obrakmann replied:
Oh, I thought we were using it for the xdg shortcuts, but it seems the Makefile writes its own copy.

obrakmann commented on launch-game.sh:
```diff
@@ -1,9 +1,36 @@
-	$ZENITY --question --title "OpenRA" --text "OpenRA has encountered a fatal error.\nLog Files are available in ~/.openra." --ok-label "Quit" --cancel-label "View FAQ" || xdg-open https://github.com/OpenRA/OpenRA/wiki/FAQ
+	if which zenity
+	then
+		TITLE=$(zenity --forms --add-combo="" --combo-values="Red Alert|Tiberian Dawn|Dune 2000|Tiberian Sun" --text "Select mod" --title="" || echo "cancel")
```
How would this handle custom mods?

pchote replied:
It doesn't and isn't supposed to.  That is [launch-mod.sh](https://github.com/OpenRA/OpenRAModTemplate/blob/master/launch-mod.sh)'s job.

obrakmann reviewed (approved):
👍 then

atlimit8 commented on launch-game.sh:
```diff
@@ -1,9 +1,36 @@
 then
-	ZENITY=`which zenity` || echo "OpenRA needs zenity installed to display a graphical error dialog. See ~/.openra. for log files."
-	$ZENITY --question --title "OpenRA" --text "OpenRA has encountered a fatal error.\nLog Files are available in ~/.openra." --ok-label "Quit" --cancel-label "View FAQ" || xdg-open https://github.com/OpenRA/OpenRA/wiki/FAQ
+	if which zenity
```
`if which zenity>/dev/null`
The output of `which` is just noise.

atlimit8 commented on launch-game.sh:
```diff
@@ -1,9 +1,36 @@
+# Show a crash dialog if required
+if [ $? != 0 ] && [ $? != 1 ]
+then
+	if which zenity
```
`if which zenity>/dev/null` here as well.

atlimit8 reviewed (changes requested):
Just a couple of nits.

pchote commented:
Fixed and rebased.

atlimit8 commented:
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/305ea7566986989b3242af00a7c8a3aeac4fba0d...702b2d087f6e4f43261fc5dbbe85a3e6d3e470d7)
